.github/scripts/check_output.py

```diff
@@ -0,0 +1,28 @@
+from pathlib import Path
+
+# TODO: The main.cpp file should print progress markers with a given string in them and this should look only through each of those in order and match them to expected result.
+
+if __name__ == '__main__':
+    data = Path('serial.log').read_text()
+    if "GNU GRUB" not in data:
+        print("ERROR: Didn't load GRUB.")
+        exit(1)
+    if "LOADED OS. Entering event loop." not in data:
```

Comment:
You can probably decide on a structured format for logging from the OS e.g. `SEVERITY TIME [COMPONENT]: MESSAGE` and just parse that out to see if you get any `FATAL 1/1/2025 15:56 [GRUB]: Failed to read boot tables` 

You could also have `INFO 1/1/2025 15:56 [X]: Initialised` and report and assert on the status of each component as they successfully init.